connection_tracker/conn_track/conn_tracker.cpp

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+// Windows.h needs to be the first include to prevent failures in subsequent headers.
+#include <windows.h>
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include <ebpf_api.h>
+#include <io.h>
+#include <iostream>
+#include <ip2string.h>
+#include <in6addr.h>
+#include <mstcpip.h>
+#include <string>
+#include <condition_variable>
+#include <mutex>
+#include <unordered_map>
+
+#include "conn_tracker.h"
+
+#pragma comment(lib, "ebpfapi.lib")
+#pragma comment(lib, "ntdll.lib")
+#pragma comment(lib, "ws2_32.lib")
+
+typedef struct _connection_history
+{
+    connection_tuple_t tuple;
+    bool ipv4;
+    uint64_t start_time;
+    uint64_t end_time;
+} connection_history_t;
+
+std::unordered_map<uint32_t, std::string> _protocol = {{6, "TCP"}, {17, "UDP"}};
+
+std::string
+trim(const std::string& str)
+{
+    auto start = str.find_first_not_of(' ');
+    if (start == std::string::npos)
+        return "";
+    auto end = str.find_first_of(' ', start);
+    if (end == std::string::npos)
+        return str.substr(start);
+    else {
+        return str.substr(start, end - start);
+    }
+}
+
+#define MAX_IPV4_ADDRESS_LENGTH 16
+#define MAX_IPV6_ADDRESS_LENGTH 46
+
+std::string
+ip_address_to_string(bool ipv4, const ip_address_t& ip_address)
+{
+    std::string buffer;
+    if (ipv4) {
+        buffer.resize(MAX_IPV4_ADDRESS_LENGTH);
+        in_addr addr;
+        addr.S_un.S_addr = ip_address.ipv4;
+        auto end = RtlIpv4AddressToStringA(&addr, buffer.data());
+        buffer.resize(end - buffer.data());
+    } else {
+        buffer.resize(MAX_IPV6_ADDRESS_LENGTH);
+        in_addr6 addr;
+        memcpy(addr.u.Byte, ip_address.ipv6, sizeof(ip_address.ipv6));
+        auto end = RtlIpv6AddressToStringA(&addr, buffer.data());
+        buffer.resize(end - buffer.data());
+    }
+
+    return "[" + trim(buffer) + "]";
+}
+
+extern "C"
+{
+    int
+    conn_track_history_callback(void* ctx, void* data, size_t size);
+}
+
+int
+conn_track_history_callback(void* ctx, void* data, size_t size)
+{
+    UNREFERENCED_PARAMETER(ctx);
+
+    if (size == sizeof(connection_history_t)) {
+        auto history = reinterpret_cast<connection_history_t*>(data);
+        auto source = ip_address_to_string(history->ipv4, history->tuple.src_ip) + ":" +
+                      std::to_string(htons(history->tuple.src_port));
+        auto dest = ip_address_to_string(history->ipv4, history->tuple.dst_ip) + ":" +
+                    std::to_string(htons(history->tuple.dst_port));
+        double duration = static_cast<double>(history->end_time);
+        duration -= static_cast<double>(history->start_time);
+        duration /= 1e9;
+        std::cout << source << "==>" << dest << "\t" << _protocol[history->tuple.protocol] << "\t" << duration
```

Comment:
also output compartment id and interface (call `ConvertInterfaceLuidToNameW()` to get the interface name to print)